### PR TITLE
CarmanKozenyDragForce

### PR DESCRIPTION
**Pull request by drwells**
We're making some substantial simplifications to the advection-diffusion integrators in the other branch - meanwhile lets merge some unrelated things.

A few comments:
1. It looks like `H_var` is always the advection-diffusion source term. Is this correct? If so, lets just use `adv_diff->getSourceTerm()` to access the `CellVariable` instead of passing it in.
2. Similarly, which object is responsible for the liquid fraction? We should pass that object instead of passing the variable.
3. Lets add a simple test.

@amneetb want to pick one of your students to wrap this one up? Once its in we can rebase and update the other PRs.

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by amneetb** [comment]
@drwells : Can you point to the code, so that we don't miss it?

@Raamkrishnan : can you take care of this?

File: src/IB/CarmanKozenyDragForce.cpp [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+#include "ibamr/app_namespaces.h" // IWYU pragma: keep
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBAMR
+{
+/////////////////////////////// PUBLIC //////////////////////////////////////
+CarmanKozenyDragForce::CarmanKozenyDragForce(std::string object_name,
+                                             Pointer<CellVariable<NDIM, double> > H_var,
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:60]:
```suggestion
```

File: src/IB/CarmanKozenyDragForce.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+                                             Pointer<CellVariable<NDIM, double> > lf_var,
+                                             Pointer<AdvDiffHierarchyIntegrator> adv_diff_solver,
+                                             Pointer<INSVCStaggeredHierarchyIntegrator> fluid_solver,
+                                             Pointer<Database> input_db,
+                                             bool register_for_restart)
+    : BrinkmanPenalizationStrategy(std::move(object_name), register_for_restart),
+      d_adv_diff_solver(adv_diff_solver),
+      d_fluid_solver(fluid_solver),
+      d_H_var(H_var),
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:69]:
```suggestion
```

File: src/IB/CarmanKozenyDragForce.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    // Get the interpolated density variable
+    const int rho_ins_idx = d_fluid_solver->getLinearOperatorRhoPatchDataIndex();
+
+    // Get the cell-centered viscosity patch data index. Returns mu_scratch with ghost cells filled.
+    const int mu_ins_idx = d_fluid_solver->getLinearOperatorMuPatchDataIndex();
+
+    // Ghost fill the heaviside and liquid fraction values.
+    VariableDatabase<NDIM>* var_db = VariableDatabase<NDIM>::getDatabase();
+    const int H_idx = var_db->mapVariableAndContextToIndex(d_H_var, d_adv_diff_solver->getNewContext());
+    const int H_scratch_idx = var_db->mapVariableAndContextToIndex(d_H_var, d_adv_diff_solver->getScratchContext());
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:104]:
```suggestion
    Pointer<CellVariable<NDIM, double>> H_var = d_adv_diff_solver->getSourceTerm();
    TBOX_ASSERT(H_var);
    const int H_idx = var_db->mapVariableAndContextToIndex(H_var, d_adv_diff_solver->getNewContext());
    const int H_scratch_idx = var_db->mapVariableAndContextToIndex(H_var, d_adv_diff_solver->getScratchContext());
```

File: include/ibamr/CarmanKozenyDragForce.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ * the density or the inertia scale or both for \f$C_d\f$ through input.
+ */
+class CarmanKozenyDragForce : public BrinkmanPenalizationStrategy
+{
+public:
+    /*
+     * \brief Constructor of the class.
+     */
+    CarmanKozenyDragForce(std::string object_name,
+                          SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > H_var,
```
Comment (drwells) [inline include/ibamr/CarmanKozenyDragForce.h:63]:
```suggestion
```

File: include/ibamr/CarmanKozenyDragForce.h [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+     * \brief Pointers to solvers.
+     */
+    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> d_adv_diff_solver;
+    SAMRAI::tbox::Pointer<IBAMR::INSVCStaggeredHierarchyIntegrator> d_fluid_solver;
+
+    /*!
+     * \brief Heaviside variable defining the gas-pcm interface.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > d_H_var;
+
```
Comment (drwells) [inline include/ibamr/CarmanKozenyDragForce.h:113]:
```suggestion
```

File: src/IB/CarmanKozenyDragForce.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    typedef HierarchyGhostCellInterpolation::InterpolationTransactionComponent InterpolationTransactionComponent;
+    std::vector<InterpolationTransactionComponent> phi_transaction_comps(2);
+    phi_transaction_comps[0] = InterpolationTransactionComponent(H_scratch_idx,
+                                                                 H_idx,
+                                                                 "CONSERVATIVE_LINEAR_REFINE",
+                                                                 false,
+                                                                 "CONSERVATIVE_COARSEN",
+                                                                 "LINEAR",
+                                                                 false,
+                                                                 d_adv_diff_solver->getPhysicalBcCoefs(d_H_var));
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:122]:
```suggestion
                                                                 d_adv_diff_solver->getPhysicalBcCoefs(H_var));
```

File: src/IB/CarmanKozenyDragForce.cpp [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+    // Using levelset bc for liquid fraction.
+    phi_transaction_comps[1] = InterpolationTransactionComponent(lf_scratch_idx,
+                                                                 lf_new_idx,
+                                                                 "CONSERVATIVE_LINEAR_REFINE",
+                                                                 false,
+                                                                 "CONSERVATIVE_COARSEN",
+                                                                 "LINEAR",
+                                                                 false,
+                                                                 d_adv_diff_solver->getPhysicalBcCoefs(d_H_var));
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:132]:
```suggestion
                                                                 d_adv_diff_solver->getPhysicalBcCoefs(H_var));
```

File: src/IB/CarmanKozenyDragForce.cpp [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+    // Get the interpolated density variable
+    const int rho_ins_idx = d_fluid_solver->getLinearOperatorRhoPatchDataIndex();
+
+    // Get the cell-centered viscosity patch data index. Returns mu_scratch with ghost cells filled.
+    const int mu_ins_idx = d_fluid_solver->getLinearOperatorMuPatchDataIndex();
+
+    VariableDatabase<NDIM>* var_db = VariableDatabase<NDIM>::getDatabase();
+    const int H_idx = var_db->mapVariableAndContextToIndex(d_H_var, d_adv_diff_solver->getNewContext());
+    const int H_scratch_idx = var_db->mapVariableAndContextToIndex(d_H_var, d_adv_diff_solver->getScratchContext());
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:212]: Same here

File: src/IB/CarmanKozenyDragForce.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    typedef HierarchyGhostCellInterpolation::InterpolationTransactionComponent InterpolationTransactionComponent;
+    std::vector<InterpolationTransactionComponent> phi_transaction_comps(2);
+    phi_transaction_comps[0] = InterpolationTransactionComponent(H_scratch_idx,
+                                                                 H_idx,
+                                                                 "CONSERVATIVE_LINEAR_REFINE",
+                                                                 false,
+                                                                 "CONSERVATIVE_COARSEN",
+                                                                 "LINEAR",
+                                                                 false,
+                                                                 d_adv_diff_solver->getPhysicalBcCoefs(d_H_var));
```
Comment (drwells) [inline src/IB/CarmanKozenyDragForce.cpp:229]: etc.

**Review by drwells — COMMENTED** [review]
Sure - AFAICT, based on the phase-change examples, it is always the source term. If this isn't true then we should do something else.

**Comment by Raamkrishnan** [comment]
> Sure - AFAICT, based on the phase-change examples, it is always the source term. If this isn't true then we should do something else.

H_var  is a scalar variable that is being advected using AdvDiffSemiImplicitHierarchyIntegrator class. It is not a source term. It is one of the Q_var maintained by AdvDiffSemiImplicitHierarchyIntegrator class. Another Q_var maintained is phi_var (level-set). We solve `\partial H/\partial t+ \div (H u) = H \div (u)` to get H.

**Comment by drwells** [comment]
Got it - I misread what
```cpp
        // Register H Div U term in the Heaviside equation.
        Pointer<CellVariable<NDIM, double> > F_var = new CellVariable<NDIM, double>(H_var->getName() + "_F");
        adv_diff_integrator->registerSourceTerm(F_var, true);
        Pointer<CartGridFunction> H_forcing_fcn = new HeavisideForcingFunction(
            "H_forcing_fcn", adv_diff_integrator, H_var, time_integrator->getAdvectionVelocityVariable());
        adv_diff_integrator->setSourceTermFunction(F_var, H_forcing_fcn);
        adv_diff_integrator->setSourceTerm(H_var, F_var);
```
does in one of the examples.

**Comment by drwells** [comment]
I'm happy to help get this done but ultimately the authors of the code are responsible for ensuring we meet all requirements for a new IBAMR feature (code review, passing the CI, new tests). What's the plan for wrapping this up?

**Comment by Raamkrishnan** [comment]
@drwells I think all your comments except the "test" for this class are addressed. This class requires information on levelset and liquid fraction to impose velocity on the solid. PhaseChangeHierarchyIntegrator maintains the liquid fraction. I cannot think of any simple test without including phase change classes. Phase change classes are still under review. @amneetb Do you have any suggestions for the test?

**Comment by drwells** [comment]
If you just need the liquid fraction, one possibility would be to just define the variable on your own and set its values to something reasonable - that would make a test independent of that infrastructure (also known as mocking).

**Comment by Raamkrishnan** [comment]
> If you just need the liquid fraction, one possibility would be to just define the variable on your own and set its values to something reasonable - that would make a test independent of that infrastructure (also known as mocking).

@drwells tests/phase_change/ already makes use of CarmanKozenyDragForce class. I think that should be enough.

**Comment by drwells** [comment]
We'll need that test, but each PR needs to come with its own tests and documentation - we need a test for this feature in isolation in addition to the more complex ones present on the other branch.

**Comment by amneetb** [comment]
No problem. We can add a test.

**Comment by drwells** [comment]
It looks like this is duplicated by #1626 - do you want to close this PR and put it in that one instead? I'm happy with whatever direction we go in as long as we still write tests and do code review.

**Comment by amneetb** [comment]
Yes, we can close this one.

**Pull request closed without merging**